app/attributes/fieldvalidator.cpp

```diff
@@ -0,0 +1,259 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "fieldvalidator.h"
+#include "attributedata.h"
+#include "featurelayerpair.h"
+#include "appresources.h"
+
+#include "qgsfield.h"
+#include "qgsvectorlayerutils.h"
+
+#include <QRegularExpression>
+#include <QLocale>
+
+FieldValidator::FieldValidator( QObject *parent ) :
+  QObject( parent )
+{
+}
+
+bool FieldValidator::validate( const FeatureLayerPair &pair, const FormItem &item, QString &validationMessage, ValidationMessageLevel &level )
+{
+  validationMessage = QString();
+  level = Info;
+
+  // Ignore fid field and relations
+  if ( item.type() != FormItem::Field || item.name() == QStringLiteral( "fid" ) )
+  {
+    return true;
+  }
+
+  bool valueIsValid = true;
+
+  const QgsField field = item.field();
+  QVariant value = pair.feature().attribute( item.fieldIndex() );
+
+  bool isNumeric = item.editorWidgetType() == QStringLiteral( "Range" ) || field.isNumeric();
+  if ( isNumeric )
+  {
+    valueIsValid = validateNumericField( item, value, validationMessage, level );
+  }
+  else if ( item.editorWidgetType() == QStringLiteral( "TextEdit" ) )
+  {
+    valueIsValid = validateTextField( item, value, validationMessage, level );
+  }
+  else
+  {
+    valueIsValid = validateGenericField( item, value, validationMessage, level );
+  }
+
+  if ( !valueIsValid )
+    return false;
+
+  // Continue to check hard and soft QGIS constraints
+  QStringList errors;
+
+  bool hardConstraintSatisfied = QgsVectorLayerUtils::validateAttribute( pair.layer(),  pair.feature(), item.fieldIndex(), errors, QgsFieldConstraints::ConstraintStrengthHard );
+  if ( !hardConstraintSatisfied )
+  {
+    validationMessage = constructConstraintValidationMessage( item, errors );
+    level = Error;
+    return false;
+  }
+
+  errors.clear();
+
+  bool softConstraintSatisfied = QgsVectorLayerUtils::validateAttribute( pair.layer(),  pair.feature(), item.fieldIndex(), errors, QgsFieldConstraints::ConstraintStrengthSoft );
+  if ( !softConstraintSatisfied )
+  {
+    validationMessage = constructConstraintValidationMessage( item, errors );
+    level = Warning;
+    return false;
+  }
+
+  return true;
+}
+
+bool FieldValidator::validateTextField( const FormItem &item, QVariant &value, QString &validationMessage, ValidationMessageLevel &level )
+{
+  const QgsField field = item.field();
+
+  if ( !field.convertCompatible( value ) )
+  {
+    validationMessage = Resources::Texts::Validation::genericValidationFailed;
```

Comment:
My preference would be to have the error messages inline in the code, as it is easier to read the code + they are only used once in the code anyway (in tests it is best if the expected error messages are intentionally duplicated)